### ISSUE TITLE
Meteor 1.2 - Stuck in `Client modified -- refreshing`

### ISSUE THREAD

**Issue by nathan-muir** [meteor/meteor#5074]
![stuck-refreshing](https://cloud.githubusercontent.com/assets/3112329/9571353/a75bf2ee-4fdf-11e5-9bd8-cb6e9a2faa20.gif)

I would love to provide some more detail on this one; Is there some environment flags I can add to produce some debugging output?

----

What I'm doing:
Editing a package that includes `.jsx` and `.coffee` files. Editing one or the other trips it into continuously refreshing. 

However, the browser only reloads once.

---
Update:
Heres a sample of the output with the various debugging flags enabled

https://gist.github.com/nathan-muir/21a656c53d34dac99d2c

**Comment by gabrielalmeida** [comment]
Hi, try debugging w/ those flags:

```
$ METEOR_PROFILE_REQUIRE=1 METEOR_PROFILE=1 METEOR_LOG=debug VELOCITY_DEBUG=true meteor --verbose
```

**Comment by gabrielalmeida** [comment]
Similar issue reported on #4943:

@raix commented 2 days ago:
>I'm getting an infinite "client update" sometimes - 1.2-rc.7 mac - I'm guessing it could be a watch loop?

>=> App running at: http://localhost:3000/
=> Client modified -- refreshing (x440)^C

**Comment by nathan-muir** [comment]
Added a log of it stuck in a refresh loop.

**Comment by sebakerckhof** [comment]
I experience the same issue, running with the flags you mention outputs (after I make a change):
https://gist.github.com/sebakerckhof/654132f67de993659f52

I only posted the first few iterations, but it continues in the same manner

**Comment by gabrielalmeida** [comment]
I did not managed to reproduce this. Can you share a minimal reproduction repo or provide contents of ```.meteor/packages```, ```.meteor/release``` and ```.meteor/versions```?

Guessing it's something related with packages that do use the plugin API, as @nathan-muir seems to experience this only when changing files with extensions affected by the jsx/coffee compilers.

Try performing a ```meteor reset; meteor update --packages-only``` and see if it helps someway. 
Be aware that ```meteor reset``` removes all local data, skip it if you don't want to lose your collections for ex.

**Comment by sebakerckhof** [comment]
Well, I actually experience this with my own fork of https://github.com/sebakerckhof/meteor-scss which I rewrote for 1.2. But since I'm the only one using it, it didn't think the fault would be there.

**Comment by gabrielalmeida** [comment]
@sebakerckhof Running a fresh app with your fork just works for me too

**Comment by sebakerckhof** [comment]
@gabrielalmeida try this repo: https://github.com/sebakerckhof/meteorrefreshloop and edit the file packages/custom/variables.import.scss

**Comment by stubailo** [comment]
@sebakerckhof Thanks for the reproduction! We'll look into this soon!

**Comment by glasser** [comment]
Looking into the reproduction. In order to run it, I had to delete `packages/fourseven_scss/.npm/plugin/compileScssBatch/npm-shrinkwrap.json`, which appeared to have cached references to somebody's private npm mirror at kkndedu01.barco.com.

**Comment by sebakerckhof** [comment]
Yeah, my bad, that is the private npm repo from the project I'm working on.

**Comment by glasser** [comment]
@stubailo This bug was introduced in 87b11bdab5c308935b88d3db6d99c71e80b36c38 when the firstRun variable in run-app's `_runOnce` was changed to a const and the `firstRun = false` line was removed.  `_runOnce`'s name is a bit confusing: it means "run the server once", but with client-only refreshes this can involve running the build step multiple times.  The removal of this line led the runner to not reset the project context if only client files changed, meaning that among other things local packages didn't get rebuilt.  I'll commit the fix.

**Comment by glasser** [comment]
I pushed a fix (71cf002).  @stubailo  maybe you can write a test :)

@martijnwalraven Can you review my fix and make sure I wasn't missing anything related to why you removed that `firstRun = true` line?  I assume it was just because that's a very big and complex function and it wasn't obvious that bundleApp gets called more than once.

**Comment by glasser** [comment]
(BTW this is a pretty nasty bug so you may want to put it on devel? I'm not sure what your branch management strategy is right now)

**Comment by stubailo** [comment]
I can just merge release 1.2 into devel if that helps. What would the test be?

**Comment by glasser** [comment]
test that if the first thing you do after running an app is make a client-only (eg css) change in a local package, you don't get an infinite refresh loop?  I suspect that would have happened (ie, extra packages like scss not needed)

**Comment by martijnwalraven** [comment]
Ah, I'm really sorry, thanks for fixing this! I assumed `options.firstRun` being passed into `_runOnce()` meant it would only be modified within the loop in `_fiber()`. I didn't realize `bundleApp()` could be called multiple times within one `_runOnce` invocation. Nested functions and state can be hard to follow, but I shouldn't have changed it without checking.

**Comment by glasser** [comment]
It's a really complicated function!

**Comment by bartmichu** [comment]
I'm still experiencing this issue with 1.2.1 but only after adding ```accounts-password```

**Comment by bartmichu** [comment]
Turns out it has nothing to do with ```accounts-password``` as I reverted all changes related to that package and I'm still having the same problem.
Right now I'm really having difficulties with narrowing it down but it looks like it was introduced very recently with some updates.

**Comment by bartmichu** [comment]
Argh I apologize, it was totally my fault.

**Comment by bevinhex** [comment]
I always have it while I was modifying css files, not while editing html, here is my packages
```
-> cat .meteor/versions 
accounts-base@1.2.2
accounts-password@1.1.4
accounts-ui-unstyled@1.1.8
autopublish@1.0.4
autoupdate@1.2.4
babel-compiler@5.8.24_1
babel-runtime@0.1.4
base64@1.0.4
binary-heap@1.0.4
blaze@2.1.3
blaze-html-templates@1.0.1
blaze-tools@1.0.4
boilerplate-generator@1.0.4
caching-compiler@1.0.0
caching-html-compiler@1.0.2
callback-hook@1.0.4
check@1.1.0
coffeescript@1.0.11
cosmos:browserify@0.9.3
dburles:mongo-collection-instances@0.3.4
ddp@1.2.2
ddp-client@1.2.1
ddp-common@1.2.2
ddp-rate-limiter@1.0.0
ddp-server@1.2.2
deps@1.0.9
diff-sequence@1.0.1
ecmascript@0.1.6
ecmascript-runtime@0.2.6
ejson@1.0.7
email@1.0.8
es5-shim@4.1.14
fastclick@1.0.7
flemay:less-autoprefixer@1.2.0
geojson-utils@1.0.4
hot-code-push@1.0.0
html-tools@1.0.5
htmljs@1.0.5
http@1.1.1
id-map@1.0.4
insecure@1.0.4
jquery@1.11.4
jsx@0.2.3
kadira:flow-router@2.10.1
lai:collection-extensions@0.1.4
launch-screen@1.0.4
less@2.5.1
livedata@1.0.15
localstorage@1.0.5
logging@1.0.8
markoshust:radium@0.15.3
meteor@1.1.10
meteor-base@1.0.1
meteortoys:toykit@2.2.1
minifiers@1.1.7
minimongo@1.0.10
mobile-experience@1.0.1
mobile-status-bar@1.0.6
mongo@1.1.3
mongo-id@1.0.1
msavin:mongol@1.6.2
npm-bcrypt@0.7.8_2
npm-mongo@1.4.39_1
observe-sequence@1.0.7
ordered-dict@1.0.4
promise@0.5.1
random@1.0.5
rate-limit@1.0.0
react@0.14.3
react-meteor-data@0.2.4
react-runtime@0.14.4
react-runtime-dev@0.14.4
react-runtime-prod@0.14.4
reactive-dict@1.1.3
reactive-var@1.0.6
reload@1.1.4
retry@1.0.4
routepolicy@1.0.6
semantic:ui@2.1.7
semantic:ui-data@2.1.7
service-configuration@1.0.5
session@1.1.1
sha@1.0.4
spacebars@1.0.7
spacebars-compiler@1.0.7
srp@1.0.4
standard-minifiers@1.0.2
templating@1.1.5
templating-tools@1.0.0
tracker@1.0.9
ui@1.0.8
underscore@1.0.4
url@1.0.5
webapp@1.2.3
webapp-hashing@1.0.5
```

**Comment by freshfey** [comment]
happens to me as well, similarly to @bevinhex (while editing the css)

`-> cat .meteor/versions
autopublish@1.0.4
autoupdate@1.2.4
babel-compiler@5.8.24_1
babel-runtime@0.1.4
base64@1.0.4
binary-heap@1.0.4
blaze@2.1.3
blaze-html-templates@1.0.1
blaze-tools@1.0.4
boilerplate-generator@1.0.4
caching-compiler@1.0.0
caching-html-compiler@1.0.2
callback-hook@1.0.4
check@1.1.0
ddp@1.2.2
ddp-client@1.2.1
ddp-common@1.2.2
ddp-server@1.2.2
deps@1.0.9
diff-sequence@1.0.1
ecmascript@0.1.6
ecmascript-runtime@0.2.6
ejson@1.0.7
emojione:emojione@2.1.0
es5-shim@4.1.14
fastclick@1.0.7
geojson-utils@1.0.4
hot-code-push@1.0.0
html-tools@1.0.5
htmljs@1.0.5
http@1.1.1
id-map@1.0.4
insecure@1.0.4
jquery@1.11.4
launch-screen@1.0.4
livedata@1.0.15
logging@1.0.8
meteor@1.1.10
meteor-base@1.0.1
minifiers@1.1.7
minimongo@1.0.10
mobile-experience@1.0.1
mobile-status-bar@1.0.6
mongo@1.1.3
mongo-id@1.0.1
npm-mongo@1.4.39_1
observe-sequence@1.0.7
ordered-dict@1.0.4
promise@0.5.1
random@1.0.5
reactive-dict@1.1.3
reactive-var@1.0.6
reload@1.1.4
retry@1.0.4
routepolicy@1.0.6
session@1.1.1
spacebars@1.0.7
spacebars-compiler@1.0.7
standard-minifiers@1.0.2
templating@1.1.5
templating-tools@1.0.0
tracker@1.0.9
twbs:bootstrap@3.3.6
ui@1.0.8
underscore@1.0.4
url@1.0.5
webapp@1.2.3
webapp-hashing@1.0.5`
